processing/src/main/java/org/apache/druid/segment/join/JoinVirtualColumnSplit.java

```diff
@@ -0,0 +1,161 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.segment.join;
+
+import org.apache.druid.query.filter.Filter;
+import org.apache.druid.segment.VirtualColumn;
+import org.apache.druid.segment.VirtualColumns;
+import org.apache.druid.segment.column.RowSignature;
+import org.apache.druid.segment.join.filter.JoinFilterSplit;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+public class JoinVirtualColumnSplit
+{
+  /**
+   * Split {@link VirtualColumns} into those which can have inputs which exist on the base table and those which can
+   * be deferred until after the join due to not being present in the base table and thus all null
+   */
+  public static JoinVirtualColumnSplit split(
+      RowSignature baseTableSignature,
+      VirtualColumns virtualColumns,
+      JoinFilterSplit filterSplit
+  )
+  {
+    final Comparator<VirtualColumn> sorter = Comparator.comparing(VirtualColumn::getOutputName);
+    final SortedSet<VirtualColumn> preJoinVirtualColumns = new TreeSet<>(sorter);
+    final SortedSet<VirtualColumn> postJoinVirtualColumns = new TreeSet<>(sorter);
+    final Set<String> baseColumnNames = new HashSet<>(baseTableSignature.getColumnNames());
+    final Set<String> baseFilterColumns = filterSplit.getBaseTableFilter()
+                                                     .map(Filter::getRequiredColumns)
+                                                     .orElse(Collections.emptySet());
+    final Set<String> unresolvedNames = new HashSet<>();
+
+    final Set<VirtualColumn> unresolvedSet = new HashSet<>();
+    // do a first pass to eliminate those which depend directly on the base table or are required by base table filters
+    for (VirtualColumn virtualColumn : virtualColumns.getVirtualColumns()) {
+      if (baseColumnNames.containsAll(virtualColumn.requiredColumns())) {
+        // virtual column depends on physical columns, or virtual columns which have already been added to the 'pre' set
+        preJoinVirtualColumns.add(virtualColumn);
+        baseColumnNames.add(virtualColumn.getOutputName());
```

Comment:
this section could be removed entirely since you will add all these in a follow-up block. So you take first pass and add any virtual columns and its dependencies required by base filter to preJoinVirtualColumn and baseColumnNames. 

then you take another pass and add virtual columns whose required columns are in baseColunnNames to preJoinVirtualColumn. Here too, I think, recursion would be easier to follow. Like before checking if required columns are in baseColumnNames, you can recursively check dependencies and add them to baseColumnNames and after that, you can check the root virtual column if its required columns are in baseColumnNames.

Reply:
i think its better to keep them in the first loop, the 2nd loop doesn't happen if there are no virtual columns that depend on other virtual columns and none of their inputs are missing from the base table, but has to if we move this out. This is done per segment, so not a ton i suppose, but not once per query either, so the less work we do feels better to me.

does recursion really make it that much clearer?

Reply:
to me, at least, the recursive version looks cleaner.

Reply:
instead i just basically deleted most of this logic since it wasn't really that useful. the important thing here is that virtual columns that refer to things that are only on the join are available for the post-join cursor, and this was already being split out by the join pre-filter analysis. Its fine that virtual columns that refer to non-existent inputs are on the base adapter and doesn't cause any problems or even really change anything afaict.

So now, we simply push all of the virtual columns which do not depend on join table columns into the base adapter, and then use the post-join virtual columns already computed by the pre-filter analysis for the post join cursor (which will fall back to the base adapter for the other virtual columns).

Reply:
there might still be some issues around virtual columns that depend on virtual columns that depend on join table stuff not being part of the join cursor, but I think at this point I'd rather fix that as a follow-up PR if that's cool